File: src/main/java/com/ptaf/ai/telemetry/Telemetry.java
package com.ptaf.ai.telemetry;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ObjectNode;
import com.ptaf.ai.BuildInfo;

import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.StandardOpenOption;
import java.time.Instant;

/**
 * Phase 4: append-only JSON lines to {@code target/ai-telemetry.jsonl} (usage + operation + version).
 */
public final class Telemetry {

    private static final ObjectMapper JSON = new ObjectMapper();

    private Telemetry() {
    }

    public static void logGeminiResponse(String operation, String model, String responseBody) {
        try {
            JsonNode root = JSON.readTree(responseBody);
            ObjectNode line = JSON.createObjectNode();
            line.put("ts", Instant.now().toString());
            line.put("moduleVersion", BuildInfo.VERSION);
            line.put("operation", operation != null ? operation : "unknown");
            line.put("model", model);
            JsonNode usage = root.get("usageMetadata");
            if (usage != null && !usage.isMissingNode()) {
                line.set("usageMetadata", usage);
            } else {
                line.putNull("usageMetadata");
            }
            Path target = Path.of("target");
            Files.createDirectories(target);
            Path log = target.resolve("ai-telemetry.jsonl");
            Files.writeString(
                    log,
                    JSON.writeValueAsString(line) + "\n",
                    StandardCharsets.UTF_8,
                    StandardOpenOption.CREATE,
                    StandardOpenOption.APPEND
            );
        } catch (Exception ignored) {
            // never fail generation on telemetry
        }
    }
}

File: python/ptaf/ai/telemetry/telemetry.py
from __future__ import annotations

import json
from datetime import datetime, timezone
from pathlib import Path

from ptaf.ai.build_info import VERSION


def log_gemini_response(operation: str, model: str, response_body: str) -> None:
    try:
        parsed = json.loads(response_body)
        line = {
            "ts": datetime.now(timezone.utc).isoformat(),
            "moduleVersion": VERSION,
            "operation": operation or "unknown",
            "model": model,
            "usageMetadata": parsed.get("usageMetadata"),
        }
        target = Path("target")
        target.mkdir(parents=True, exist_ok=True)
        log_path = target / "ai-telemetry.jsonl"
        with log_path.open("a", encoding="utf-8") as handle:
            handle.write(json.dumps(line) + "\n")
    except Exception:
        pass
